--- devtree.c	2014-05-16 21:20:12.000000000 +0500
+++ devtree_modified.c	2014-05-20 13:52:34.000000000 +0500
@@ -80,7 +80,7 @@
	 */
	struct device_node *cpu, *cpus;
	u32 i, j, cpuidx = 1;
-	u32 mpidr = is_smp() ? read_cpuid_mpidr() & MPIDR_HWID_BITMASK : 0;
+	u32 mpidr = read_cpuid_mpidr() & MPIDR_HWID_BITMASK;

	u32 tmp_map[NR_CPUS] = { [0 ... NR_CPUS-1] = UINT_MAX };
	bool bootcpu_valid = false;
